File: src/main/java/kanti/sl/arguments/MutableStateArgumentImpl.java
package kanti.sl.arguments;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class MutableStateArgumentImpl implements MutableStateArgument {

	@NotNull
	private String key;
	@Nullable
	private Object value;

	public MutableStateArgumentImpl() {
		key = "";
	}

	public MutableStateArgumentImpl(@NotNull String key) {
		this(key, null);
	}

	public MutableStateArgumentImpl(@NotNull String key, @Nullable Object value) {
		this.key = key;
		this.value = value;
	}

	@Override
	public void setKey(@NotNull String key) {
		this.key = key;
	}

	@Override
	public void setValue(@NotNull Object value) {
		TypeChecker.Type type = TypeChecker.getType(value);
		this.value = TypeChecker.convert(value, type);
	}

	@NotNull
	@Override
	public String getKey() {
		return key;
	}

	@NotNull
	@Override
	public Object getValue() {
		if (value == null)
			throw new IllegalStateException("State not initialized");
		return value;
	}

}

File: src/main/kotlin/kanti/sl/arguments/MutableStateArgumentImpl.kt
package kanti.sl.arguments

class MutableStateArgumentImpl @JvmOverloads constructor(
	override var name: String = "",
	value: Any? = null
) : MutableStateArgument {

	private var _value: Any? = value
	override var value: Any
		get() {
			val v = _value
			check(v != null) {
				"Value not initialized!"
			}
			return v
		}
		set(value) {
			val checkedType = TypeChecker.getType(value)
			_value = TypeChecker.convert(value, checkedType)
		}

}